src/main/java/com/google/devtools/build/lib/syntax/EmptyExpression.java

```diff
@@ -0,0 +1,27 @@
+package com.google.devtools.build.lib.syntax;
+
+/** An expression without content. */
```

Comment:
The reason why I chose to use a null object instead of just passing null is that it is a precondition that buildTagFilter is not null in [TargetPatternPhaseValue](https://github.com/bazelbuild/bazel/blob/1fe12e26a4b3d898326385a05b15b9a4408d5454/src/main/java/com/google/devtools/build/lib/skyframe/TargetPatternPhaseValue.java#L225). Using null would invalidate this. When removing the precondition this results in many test failures. 
Is there any chance you might reconsider? If so do you have a suggestion for the name if you are not pleased with empty?

Reply:
We absolutely cannot add a new expression type to the language for this: it would be adding complexity to a widely used lower-level package just to support a single esoteric use in a higher-level package, and that is nearly always a mistake. Every existing client of the lower-level package would be immediately broken because their switch(stmt.kind()) statements would become non-exhaustive. So they would have add a case for EmptyExpression, but what should it do? What does empty expression even mean? Where does it come from? Why isn't it mentioned in the language spec? And so on.

(One could avoid null by using some trivially true expression such as "x or not x", but that's still a hack, so let's not do that.)

The obvious solution is simply to remove the non-null precondition check, as it no longer applies. Make clear that this is intentional by annotating the relevant parameters, fields, and accessor methods with @Nullable. Follow the flow of values as far as it goes so that we don't pass null to a function that requires non-null. This particular value doesn't flow very far.